applications/luci-app-attendedsysupgrade/files/attendedsysupgrade.htm

```diff
@@ -0,0 +1,342 @@
+	request_dict = {}
+	request_dict.version = latest_version;
+	request_dict.board = data.board_name
+	server_request(request_dict, "/image-request", image_request_handler)
+}
+
+function image_request_handler(response) {
+	if (response.status === 400) {
+		response_content = JSON.parse(response.responseText)
+		update_error(response_content.error)
+	} else if (response.status === 500) { 
+		image_request_500()
+	} else if (response.status === 503) {
+		update_error("please wait. server overloaded")
+		// handle overload
+		setTimeout(image_request, 30000)
+	} else if (response.status === 201) {
+		response_content = JSON.parse(response.responseText)	
+		if(response_content.queue != undefined) {
+			// in queue
+			update_info("please wait. you are in queue position " + response_content.queue)
+			console.log("queued")
+		} else {
+			update_info("imagebuilder not ready, please wait")
+			console.log("setting up imagebuilder")
+		}
+		setTimeout(image_request, 5000)
+	} else if (response.status === 206) {
+		// building
+		console.log("building")
+		update_info("building image")
+		setTimeout(image_request, 5000)
+	} else if (response.status === 200) {
+		// ready to download
+		response_content = JSON.parse(response.responseText)
+		update_info("image created")
+		document.getElementById("update_button").value = "download and flash"
+		document.getElementById("update_button").onclick = function() {download_image(response_content.url); }
+	}
+}
+
+
+// uploads received blob data to the server using cgi-io
+function upload_image(blob) {
+	var upload_request = new XMLHttpRequest();
+	var form_data  = new FormData();
+
+	form_data.append("sessionid", data.ubus_rpc_session)
+	form_data.append("filename", "/tmp/sysupgrade.bin")
+	form_data.append("filemode", 755) // insecure?
+	form_data.append("filedata", blob)
+	
+	upload_request.addEventListener('load', function(event) {
+		// this checksum should be parsed
+		document.getElementById("update_info").innerHTML = "flashing... please wait" // show fancy indicator http://www.ajaxload.info/
+		ubus_request("attendedsysupgrade", "sysupgrade", '{  }', 'done');
+  	});
+
+  	upload_request.addEventListener('error', function(event) {
+		document.getElementById("update_info").innerHTML = "uploading failed, please retry"
+  	});
+
+  	upload_request.open('POST', origin + '/cgi-bin/cgi-upload');
+  	upload_request.send(form_data);
+}
+
+// download image from server once the url was received by image_request
+function download_image(url) {
+	console.log("download_image")
+	document.getElementById("update_button").value = "flashing..."
+	document.getElementById("update_button").disabled = true;
+	var download_request = new XMLHttpRequest();
+	download_request.open("GET", url);
+	download_request.responseType = "arraybuffer";
+
+	download_request.onload = function () {
+		if (this.status === 200) {
+			var blob = new Blob([download_request.response], {type: "text/plain; charset=x-user-defined"});
```

Comment:
I'm far from knowing anything about best-practices when handling binaries in JavaScript, but using `text/plain` rather than `application/octet-stream` seems odd to me